bpf/lib/overloadable_xdp.h

```diff
@@ -239,6 +239,32 @@ ctx_set_encap_info(struct xdp_md *ctx, __u32 src_ip, __be16 src_port,
 
 	return CTX_ACT_REDIRECT;
 }
+
+static __always_inline __maybe_unused int
+ctx_set_tunnel_opt(struct xdp_md *ctx, void *opt, __u32 opt_len)
+{
+	const __u32 geneve_off = ETH_HLEN + sizeof(struct iphdr) + sizeof(struct udphdr);
+	struct genevehdr geneve;
+
+	/* add free space after GENEVE header: */
+	if (ctx_adjust_hroom(ctx, opt_len, BPF_ADJ_ROOM_MAC, ctx_adjust_hroom_flags()) < 0)
+		return DROP_INVALID;
+
+	/* write the options */
+	if (ctx_store_bytes(ctx, geneve_off + sizeof(geneve), opt, opt_len, 0) < 0)
+		return DROP_WRITE_ERROR;
+
+	/* update the options length in the GENEVE header: */
+	if (ctx_load_bytes(ctx, geneve_off, &geneve, sizeof(geneve)) < 0)
+		return DROP_INVALID;
+
+	geneve.opt_len += (__u8)(opt_len >> 2);
```

Comment:
I guess opt_len must be a multiple of 4 or something. Can we `build_bug_on` or similar? Or document that somewhere.

Reply:
To a certain extent I would hope that callers know what data format they need to follow for such an option (as it also needs to work for the TC variant of `set_tunnel_opt()`), and get to keep the pieces if they don't.

But yep, a build_bug() should be doable - will queue that up as follow-on, thanks!